doc/neps/ufunc-overrides.rst

```diff
@@ -106,192 +115,492 @@ Returning ``NotImplemented`` to user should not happen. Moreover::
+method and objects can override Ufuncs by overriding this method (if
+they are :class:`ndarray` subclasses) or defining their own. The method
+signature is::
 
-    def __numpy_ufunc__(self, ufunc, method, i, inputs, **kwargs)
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs)
 
 Here:
 
 - *ufunc* is the ufunc object that was called. 
-- *method* is a string indicating which Ufunc method was called
-  (one of ``"__call__"``, ``"reduce"``, ``"reduceat"``,
-  ``"accumulate"``, ``"outer"``, ``"inner"``). 
-- *i* is the index of *self* in *inputs*.
+- *method* is a string indicating how the Ufunc was called, either
+  ``"__call__"`` to indicate it was called directly, or one of its
+  :ref:`methods<ufuncs.methods>`: ``"reduce"``, ``"accumulate"``,
+  ``"reduceat"``, ``"outer"``, or ``"at"``. 
 - *inputs* is a tuple of the input arguments to the ``ufunc``
-- *kwargs* are the keyword arguments passed to the function. The ``out``
-  arguments are always contained in *kwargs*, how positional variables
-  are passed is discussed below.
-
-The ufunc's arguments are first normalized into a tuple of input data
-(``inputs``), and dict of keyword arguments. If there are output
-arguments they are handled as follows:
-
-- One positional output variable x is passed in the kwargs dict as ``out :
-  x``.
-- Multiple positional output variables ``x0, x1, ...`` are passed as a tuple
-  in the kwargs dict as ``out : (x0, x1, ...)``.
-- Keyword output variables like ``out = x`` and ``out = (x0, x1, ...)`` are
-  passed unchanged to the kwargs dict like ``out : x`` and ``out : (x0, x1,
-  ...)`` respectively.
-- Combinations of positional and keyword output variables are not
-  supported.
+- *kwargs* contains any optional or keyword arguments passed to the
+  function. This includes any ``out`` arguments, which are always
+  contained in a tuple.
+
+Hence, the arguments are normalized: only the required input arguments
+(``inputs``) are passed on as positional arguments, all the others are
+passed on as a dict of keyword arguments (``kwargs``). In particular, if
+there are output arguments, positional are otherwise, that are not
+:obj:`None`, they are passed on as a tuple in the ``out`` keyword
+argument.
 
 The function dispatch proceeds as follows:
 
-- If one of the input arguments implements ``__numpy_ufunc__`` it is
-  executed instead of the Ufunc.
+- If one of the input or output arguments implements
+  ``__array_ufunc__``, it is executed instead of the ufunc.
 
-- If more than one of the input arguments implements ``__numpy_ufunc__``,
+- If more than one of the input arguments implements ``__array_ufunc__``,
   they are tried in the following order: subclasses before superclasses,
-  otherwise left to right.  The first ``__numpy_ufunc__`` method returning
-  something else than ``NotImplemented`` determines the return value of
-  the Ufunc.
+  inputs before outputs, otherwise left to right.
```

Comment:
`np.add(ndarray, ndarray, out=xarray)`? (Or sparse array, or dask array, etc.)

Hopefully the *semantics* of the operation shouldn't change because of the presence of `out=` (that'd be a little to matlab-y for my taste ;-)), but the main function of `__numpy_ufunc__` is find someone who knows how to iterate over all the containers involved, and being able to iterate over the output container is important.

For the same reason I think it'd probably be nice to include `where=` in the dispatch, but we can always fix that in the future once it comes up.